docs/codeql/codeql-language-guides/customizing-library-models-for-go.rst

```diff
@@ -0,0 +1,413 @@
+
+.. code-block:: go
+
+    func Tainted(db *sql.DB, name string) {
+        stmt, err := db.Prepare("SELECT * FROM users WHERE name = " + name) // The argument to this method is a SQL injection sink.
+        ...
+    }
+
+We need to add a tuple to the ``sinkModel``\(namespace, type, subtypes, name, signature, ext, input, kind, provenance) extensible predicate by updating a data extension file.
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/go-all
+         extensible: sinkModel
+       data:
+         - ["database/sql", "DB", False, "Prepare", "(string)", "", "Argument[0]", "sql-injection", "manual"]
+
+Since we want to add a new sink, we need to add a tuple to the ``sinkModel`` extensible predicate.
+The first five values identify the callable (in this case a method) to be modeled as a sink.
+
+- The first value ``database/sql`` is the package name.
+- The second value ``DB`` is the name of the type that the method is associated with.
+- The third value ``False`` is a flag that indicates whether or not the sink also applies to all overrides of the method.
+- The fourth value ``Prepare`` is the method name. Constructors are named after the class.
+- The fifth value ``(string)`` is the method input type signature. This value is often excluded and is simply set to an empty string since Go does not allow for a given type to have multiple methods with the same type.
+
+The sixth value should be left empty and is out of scope for this documentation.
+The remaining values are used to define the ``access path``, the ``kind``, and the ``provenance`` (origin) of the sink.
+
+- The seventh value ``Argument[0]`` is the ``access path`` to the first argument passed to the method, which means that this is the location of the sink.
+- The eighth value ``sql-injection`` is the kind of the sink. The sink kind is used to define the queries where the sink is in scope. In this case - the SQL injection queries.
+- The ninth value ``manual`` is the provenance of the sink, which is used to identify the origin of the sink.
+
+Example: Taint source from the ``net`` package
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This example shows how the Go query pack models the return value from the ``Listen`` method as a ``remote`` source.
+This is the ``Listen`` function which is located in the ``net`` package.
+
+.. code-block:: go
+
+    func Tainted() {
+        ln, err := net.Listen("tcp", ":8080") // The return value of this method is a remote source.
+        ...
+    }
+
+
+We need to add a tuple to the ``sourceModel``\(namespace, type, subtypes, name, signature, ext, output, kind, provenance) extensible predicate by updating a data extension file.
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/go-all
+         extensible: sourceModel
+       data:
+         - ["net", "", False, "Listen", "(string,string)", "", "ReturnValue", "remote", "manual"]
```

Comment:
I am in the middle of adapting the C# docs. There are a few things which still need re-writing.

Reply:
The example that had been used in the C# example was the API for opening a socket. As I understand it, `Listen` is roughly the equivalent in Go, and would make sense as a remote flow source.

The wording from the C# documentation mentioned it being from a existing model, because sockets were already modeled in C#.

I am fine opening a small PR to add `Listen` as a remote flow source if we think that makes sense. Alternatively, I can reword this documentation to (1) remove the wording about being an existing model or (2) change it to a different API if you have a suggestion for a better function to highlight.

Reply:
Ah, that makes sense. I recommend you talk about this model instead.

```
- ["net/http", "Request", True, "FormValue", "", "", "ReturnValue", "remote", "manual"]
```

It is a classic "yes, that definitely is a source of untrusted data" that we tend to use for Go. Go is used a lot for dealing with http requests.